## Initialization
# Load required packages
suppressPackageStartupMessages(library(tidytree, quietly = T))
suppressPackageStartupMessages(library(treeio, quietly = T))
suppressPackageStartupMessages(library(tidyverse, quietly = T))
suppressPackageStartupMessages(library(ggtree, quietly = T))
suppressPackageStartupMessages(library(jsonlite, quietly = T))
if (require(showtext, quietly = T)) {
  showtext_auto()
  if (interactive())
    showtext_opts(dpi = 100)
  else
    showtext_opts(dpi = 300)
}

## Input
if (interactive()) {
  setwd("/project/noujdine_61/kdeweese/latissima/genome_stats")
  line_args <- c(
    "busco_summaries/eukaryota_odb10",
    "s-latissima-genome/species_table.txt",
    "s-latissima-genome/"
  )
} else if (length(commandArgs(trailingOnly = T)) == 3) {
  line_args <- commandArgs(trailingOnly = T)
} else {
  stop("3 positional arguments expected.")
}
wd <- line_args[1]
assembly_file <- line_args[2]
outdir <- line_args[3]
busco_script <- "busco_figure.R"
tree_script <- "prune_tree.R"
quast_script <- "quast.R"

# Output
# Split lineage from working directory
lineage <- unlist(strsplit(wd, "/"))[2]
extens <- c("png", "tiff")
if (grepl("euk", lineage)) {
  fig_lab <- "FS1_"
} else if (grepl("strame", lineage)) {
  fig_lab <- "F2_"
} else {
  fig_lab <- ""
}
busc_plot_file <- paste0(fig_lab, "busco_", lineage, ".", extens)

# Prepend working directory to BUSCO script (if exists)
if (dir.exists(wd)) {
  busco_script <- paste(wd, busco_script, sep = "/")
}
# Prepend output directory to script and plot filenames (if it exists)
if (dir.exists(outdir)) {
  tree_script <- paste(outdir, tree_script, sep = "/")
  quast_script <- paste(outdir, quast_script, sep = "/")
  busc_plot_file <- paste0(outdir, busc_plot_file)
}

## Functions
# Labels for figure legend
lbs <- c(
  "C" = "Complete",
  "S" = "Complete and single-copy",
  "D" = "Complete and duplicated",
  "F" = "Fragmented",
  "M" = "Missing"
)
lbs_rev <- names(lbs)
names(lbs_rev) <- unname(lbs)
color_names <- unname(lbs[2:5])
# Format species Latin name
formatSpc <- function(spc) {
  spc_f <- gsub("_", " ", spc)
  # Converts Ectocarpus siliculosus to Ectocarpus sp. Ec32
  spc_f <-
    gsub("Ectocarpus siliculosus", "Ectocarpus sp. Ec32", spc_f)
  spc_f <- str_wrap(spc_f, width = 20)
  return(spc_f)
}
# Find JSON summary files for each species generated by BUSCO
getBuscoJsons <- function(species_tab) {
  busc_ids <- pull(species_tab, BUSCO_id, Species)
  ptn <- paste0(busc_ids, "$")
  ptns <- paste(ptn, collapse = "|")
  busc_res_dirs <- list.files(pattern = ptns)
  busc_res_dirs <- sapply(busc_ids, grep, busc_res_dirs, value = T)
  busc_df <-
    tibble(dir = busc_res_dirs, Species = names(busc_res_dirs))
  busc_json_files <- busc_df %>%
    rowwise() %>%
    filter(dir.exists(dir)) %>%
    mutate(json_files = list.files(
      path = dir,
      pattern = "sho.*\\.json",
      full.names = T
    )) %>%
    ungroup() %>%
    pull(json_files, Species)
  return(busc_json_files)
}
# Parse BUSCO summary JSON file
parseBusco <- function(busc_json_file) {
  busc_data <- read_json(busc_json_file, simplifyVector = T)
  busc_res <- busc_data$results %>%
    unlist() %>%
    as_tibble(rownames = "stat") %>%
    mutate(
      stat = gsub("Single copy", "Complete and single-copy", stat),
      stat = gsub("Multi copy", "Complete and duplicated", stat)
    )
  return(busc_res)
}


## Data wrangling
sourced <- T
# Source phylogenetic tree-pruning R script
print(paste("Sourcing", tree_script))
source(tree_script)
# Source QUAST output parsing R script
print(paste("Sourcing", quast_script))
source(quast_script)
# Source BUSCO-generated R script
print(paste("Sourcing", busco_script))
suppressWarnings(source(busco_script))
my_colors <- setNames(my_colors, color_names)
# Read table of species names and FASTA file names
species_tab <-
  read.table(assembly_file,
             sep = "\t",
             fill = NA,
             header = F)
colnames(species_tab)[1:2] <- c("Species", "Assembly")
# Match species table names to BUSCO IDs and tree tip labels
species_tab <- species_tab %>%
  mutate(
    BUSCO_id = paste(basename(tools::file_path_sans_ext(Assembly)),
                     lineage, sep = "_"),
    # Add underscores to match tree species names
    Species = gsub(" ", "_", Species)
  )
tree_spc <-
  sapply(tip.label(tp), grep, species_tab$Species, value = T)
tree_spc <- setNames(names(tree_spc), unname(tree_spc))
species_dict <- species_tab %>%
  mutate(Species = tree_spc[Species]) %>%
  pull(Species, BUSCO_id)

# Read in BUSCO summaries
busc_json_files <- getBuscoJsons(species_tab)
busc_res <- sapply(busc_json_files, parseBusco, simplify = F)
busc_res <- busc_res %>%
  bind_rows(.id = "Species")
busc_res_filt <- busc_res %>%
  # Keep only BUSCO gene reporting and convert values to numbers
  filter(grepl(paste(lbs, collapse = "|"), stat)) %>%
  mutate(value = as.numeric(value)) %>%
  # Separate gene counts and percentages into two columns
  separate_wider_regex(stat,
                       patterns = c(Category = ".*", "\\s",
                                    Value_Type = "per.*|BUS.*")) %>% 
  pivot_wider(names_from = Value_Type, values_from = value) %>%
  # Create one-line summary labels for each species
  mutate(
    C = lbs_rev[Category],
    Label = paste(C, BUSCOs, sep = ":"),
    Label = case_when(
      C == "S" ~ paste0("[", Label),
      C == "D" ~ paste0(Label, "]"),
      .default = Label
    )
  ) %>%
  group_by(Species) %>%
  mutate(Label = case_match(C, "S" ~ paste(Label, collapse = ", ")),
         Label = gsub(", \\[", " \\[", Label)) %>%
  ungroup() %>%
  # Sort BUSCO result categories by desired order in bars
  mutate(Category = factor(Category, levels = unname(lbs))) %>%
  # Remove "Complete" category (already represented by "S" and "D")
  filter(Category != "Complete") %>%
  # Convert Species to factor ordered by tree tip labels
  mutate(
    Species = gsub(" ", "_", Species),
    Species = tree_spc[Species],
    Species = factor(Species, levels = tip.label(tp))
  )

df2 <- quast_df %>%
  # Convert Species to factor ordered by tree tip labels
  mutate(
    Species = gsub(" ", "_", Species),
    Species = tree_spc[Species],
    Species = factor(Species, levels = tip.label(tp))
  ) %>%
  # Pivot wider for plot
  pivot_wider(id_cols = Species,
              names_from = Stat,
              values_from = Assembly)

# Format plot titles and axis labels
lin_split <- unlist(strsplit(lineage, "_"))
lin <- tools::toTitleCase(lin_split[1])
odb_v <- toupper(lin_split[2])
x_lab1 <- paste(lin, odb_v, "BUSCOs")
x_lab2 <- paste0("(n = ", total_buscos, ")")
x_lab <- paste(x_lab1, x_lab2, sep = "\n")

# Plotting
# Bar plot of BUSCO score breakdown
busc_plot <-
  ggplot(data = busc_res_filt,
         mapping = aes(x = percentage, y = Species)) +
  geom_col(mapping = aes(fill = Category),
           position = position_stack(reverse = T)) +
  geom_text(mapping = aes(label = Label), x = 5, hjust = 0, na.rm = T) +
  scale_fill_manual(
    name = NULL,
    values = my_colors,
    labels = scales::label_wrap(width = 15)
  ) +
  # guides(fill = guide_legend(ncol = 2)) +
  scale_x_continuous(
    name = x_lab,
    labels = scales::label_percent(scale = 1),
    expand = expansion(mult = 0, add = c(0, 2))
  ) +
  scale_y_discrete(labels = as_labeller(formatSpc)) +
  coord_cartesian(clip = "off") +
  theme_classic2() +
  theme(
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks.y = element_blank()
  )
tree_plot <- ggtree(tp) +
  geom_tiplab(aes(label = f_labs), align = T, fontface = "italic") +
  geom_treescale() +
  hexpand(0.5) +
  scale_x_continuous(expand = expansion(mult = 0, add = 0)) +
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.15), add = 0)) +
  coord_cartesian(clip = "off")
n_plot <- ggplot(data = df2,
                 mapping = aes(x = `# N's per 100 kbp`,
                               y = Species)) +
  geom_col() +
  scale_x_continuous(label = scales::label_comma(),
                     expand = expansion(mult = 0, add = c(0, 2))) +
  coord_cartesian(clip = "off") +
  theme_classic2() +
  theme(
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks.y = element_blank()
  )
pp <- ggarrange(
  n_plot,
  busc_plot,
  nrow = 1,
  legend = "right",
  legend.grob = get_legend(busc_plot),
  align = "h"
)
p <-
  ggarrange(tree_plot,
            pp,
            widths = c(1, 2),
            nrow = 1,
            align = "h")
# Save plot with message to user
print(paste("Saving pretty BUSCO plot of", lineage, "to", busc_plot_file))
h <- 5
w <- h * 2.5
sapply(
  busc_plot_file,
  ggsave,
  plot = p,
  bg = "white",
  width = w,
  height = h,
  simplify = F
)
